File: ruby/union_find.rb
class UFTree
  class Node
    attr_accessor :parent, :rank

    def initialize parent, rank=1
      @parent = parent
      @rank = rank
    end
  end

  def initialize size
    @nodes = Array.new(size)
    @sizes = Array.new(size)
    (0...size).each do |i|
      @nodes[i] = Node.new(i)
      @sizes[i] = 1
    end
  end

  def unite a, b
    a = root(a)
    b = root(b)

    return false if a == b

    if @nodes[a].rank < @nodes[b].rank
      @sizes[@nodes[a].parent] += @sizes[@nodes[b].parent]
      @nodes[b].parent = @nodes[a].parent
    else
      @sizes[@nodes[b].parent] += @sizes[@nodes[a].parent]
      @nodes[a].parent = @nodes[b].parent
      @nodes[a].rank += 1 if @nodes[a].rank == @nodes[b].rank
    end

    true
  end

  def same? a, b
    root(a) == root(b)
  end

  def size i
    @sizes[root(i)]
  end

  private

  def root i
    if @nodes[i].parent == i
      i
    else
      @nodes[i].parent = root(@nodes[i].parent)
    end
  end
end

File: rust/union_find.rs
struct UnionFind {
    tree: Vec<usize>,
    rank: Vec<usize>,
}

impl UnionFind {
    fn new(n: usize) -> UnionFind {
        UnionFind {
            tree : (0..n).collect::<Vec<_>>(),
            rank : vec![0;n],
        }
    }

    fn parent(&mut self, x: usize) -> usize {
        if x == self.tree[x] {
            x
        } else {
            let p = self.tree[x];
            let q = self.parent(p);
            self.tree[x] = q;
            q
        }
    }

    fn is_same(&mut self, a: usize, b: usize) -> bool {
        self.parent(a) == self.parent(b)
    }

    fn unite(&mut self, a: usize, b: usize) -> bool {
        let a = self.parent(a);
        let b = self.parent(b);
        if a == b {
            return false;
        }

        if self.rank[a] > self.rank[b] {
            self.tree[b] = a;
        } else {
            self.tree[a] = b;
            if self.rank[a] == self.rank[b] {
                self.rank[b] += 1;
            }
        }
        true
    }
}